File: nim/grade-school/grade_school.nim
import std/sequtils
import std/algorithm
# import sugar

type
  Student* = tuple[name: string, grade: int]
  School* = object
    students*: seq[Student]

proc get_names(studs: seq[Student]): seq[string] =
    var names: seq[string]
    for s in studs: 
        names.add(s.name)
    names

proc roster*(sch: School): seq[string] = 
    let sorted_school = sch.students.sortedByIt((it.grade, it.name))
    return get_names(sorted_school)

proc grade*(sch: School, g: int): seq[string] =
    let students_in_grade: seq[Student] = sch.students.filterIt(it.grade == g).sorted()
    get_names(students_in_grade)

proc add_student*(sch: var School, n: string, g: int) =
    let names = get_names(sch.students)
    if names.filterIt(it == n) != @[]:
        var e: ref ValueError
        new(e)
        raise e
    var student: Student
    student.name = n
    student.grade = g
    sch.students.add(student)

File: scala/grade-school/src/main/scala/GradeSchool.scala
class School {
  type DB = Map[Int, Seq[String]]

  private var school: DB = Map.empty

  def add(name: String, g: Int) = 
    {
      if (school.contains(g)) {
        val newSeq: Seq[String] = school(g) :+ name
        school(g) = newSeq
      }
      else school(g) = Seq(name)
    }

  def db: DB = school

  def grade(g: Int): Seq[String] =     
       if (school.contains(g)) school(g) else Seq()

  def sorted: DB = school
}

